fuse/file.go

```diff
@@ -3,103 +3,370 @@ package fuse
+		// including size only when it's been retreived previously by opening the file.
+		a.Size = f.readSize
+	}
+}
+
+func getFileMode(entry plugin.Entry) os.FileMode {
+	var mode os.FileMode
+	if plugin.WriteAction().IsSupportedOn(entry) {
+		mode |= 0220
+	}
+	if plugin.ReadAction().IsSupportedOn(entry) ||
+		plugin.StreamAction().IsSupportedOn(entry) {
+		mode |= 0440
+	}
+	return mode
+}
+
+var _ = fs.NodeOpener(&file{})
+
+// Open an entry for reading or writing. Several patterns exist for how to interact with entries.
+// - An entry that only supports Read can only be opened ReadOnly.
+// - An entry that supports both Read and Write can be opened in any mode.
+// - An entry that only supports Write can only be opened WriteOnly.
+//
+// When writing and flushing a file, we may call Read on the entry (if it supports Read) even if
+// opened WriteOnly. That only happens when performing a partial write of a *file-like* entry.
 func (f *file) Open(ctx context.Context, req *fuse.OpenRequest, resp *fuse.OpenResponse) (fs.Handle, error) {
-	activity.Record(ctx, "FUSE: Open %v", f)
-
-	// Check for an updated entry in case it has static state.
-	updatedEntry, err := f.refind(ctx)
-	if err != nil {
-		activity.Warnf(ctx, "FUSE: Open errored %v, %v", f, err)
-		return nil, err
-	}
-
-	var fh fileHandle
-	fh.id = f.String()
-
-	// Initiate content request and return a channel providing the results.
-	if plugin.ReadAction().IsSupportedOn(updatedEntry) {
-		activity.Record(ctx, "FUSE: Opened %v", f)
-		if attr := plugin.Attributes(updatedEntry); !attr.HasSize() {
-			// The entry's content size is unknown so open the file in direct
-			// IO mode. This enables FUSE to still read the entry's content so
-			// that built-in tools like cat and grep still work.
-			resp.Flags |= fuse.OpenDirectIO
+	f.mux.Lock()
+	defer f.mux.Unlock()
+	activity.Record(ctx, "FUSE: Open %v: %+v", f, *req)
+
+	if !f.useLocalState() {
+		// Check for an updated entry in case it has static state.
+		entry, err := f.refind(ctx)
+		if err != nil {
+			activity.Warnf(ctx, "FUSE: Open errored %v, %v", f, err)
+			return nil, err
 		}
-		fh.r = updatedEntry
+		f.entry = entry
+	}
+
+	readable := plugin.ReadAction().IsSupportedOn(f.entry)
+	writable := plugin.WriteAction().IsSupportedOn(f.entry)
+	switch {
+	case req.Flags.IsReadOnly() && !readable:
+		activity.Warnf(ctx, "FUSE: Open read-only unsupported on %v", f)
+		return nil, fuse.ENOTSUP
+	case req.Flags.IsWriteOnly() && !writable:
+		activity.Warnf(ctx, "FUSE: Open write-only unsupported on %v", f)
+		return nil, fuse.ENOTSUP
+	case req.Flags.IsReadWrite() && (!readable || !writable):
+		activity.Warnf(ctx, "FUSE: Open read-write unsupported on %v", f)
+		return nil, fuse.ENOTSUP
 	}
 
-	if plugin.WriteAction().IsSupportedOn(updatedEntry) {
-		fh.w = updatedEntry.(plugin.Writable)
+	if !f.isFileLike() && req.Flags.IsReadWrite() {
+		// Error ReadWrite on non-file-like entries because it probably won't work well.
```

Comment:
So with this code, if I have a non-file-like entry that implements both Read/Write, and I do something like `vim <entry>`, I'm going to see what's returned by read b/c vim opens ReadOnly. Then I do some edits and write what I see via `:w`. That's going to be written because vim opens WriteOnly (but what's written may also include the Read data, which can cause some weird issues to happen). I don't think there's anything we can do to prevent that, so might be worth adding a warning for users to be careful when editing non-file-like entries (and preferably that they _shouldn't_ edit non-file-like entries).

Reply:
Editing with vim will work, it just may do something you don't expect. Not sure what to say beyond what's already in the docs.

The comment here is because if you have a ReadWrite handle, you expect what you write to be directly reflected in read operations, and things like `seek` to make sense. They won't, because reads and writes are not symmetric for non-file-like entries.

Reply:
Oh, the code makes sense. I only commented what I said here b/c that's when I thought of it.

See https://github.com/puppetlabs/wash/pull/664/files#r362674541 re: where to add a note about unexpected things w/ editors.